hnn_core/gui/_viz_manager.py

```diff
@@ -0,0 +1,577 @@
+
+import copy
+import io
+from functools import partial
+
+import matplotlib
+import matplotlib.pyplot as plt
+import numpy as np
+from IPython.display import display
+from ipywidgets import (Box, Button, Dropdown, FloatText, HBox, Label, Layout,
+                        Output, Tab, VBox, link)
+
+from hnn_core.gui._logging import logger
+from hnn_core.viz import plot_dipole
+
+
+_dpi = 96
+_fig_placeholder = 'Run simulation to add figures here.'
+
+_plot_types = [
+    'current dipole',
+    'dipole@L2',
+    'dipole@L5',
+    'dipole@agg',
+    'input histogram',
+    'spikes',
+    'PSD',
+    'spectogram',
+    'network',
+    'layer-specific dipole',
+]
+
+_spectrogram_color_maps = [
+    "viridis",
+    "plasma",
+    "inferno",
+    "magma",
+    "cividis",
+]
+
+
+fig_templates = {
+    "2row x 1col (1:3)": {
+        "kwargs":
+        "gridspec_kw={\"height_ratios\":[1,3]}",
+        "mosaic": "00\n11",
+    },
+    "2row x 1col (1:1)": {
+        "kwargs":
+        "gridspec_kw={\"height_ratios\":[1,1]}",
+        "mosaic": "00\n11",
+    },
+    "1row x 2col (1:1)": {
+        "kwargs":
+        "gridspec_kw={\"height_ratios\":[1,1]}",
+        "mosaic": "01\n01",
+    },
+    "single figure": {
+        "kwargs": "",
+        "mosaic": "00\n00",
+    },
+    "2row x 2col (1:1)": {
+        "kwargs":
+        "gridspec_kw={\"height_ratios\":[1,1]}",
+        "mosaic": "01\n23",
+    },
+}
+
+
+def _idx2figname(idx):
+    return f"Figure {idx}"
+
+
+def _figname2idx(fname):
+    return int(fname.split(" ")[-1])
+
+
+def _update_ax(fig, ax, single_simulation, plot_type, plot_config):
+    """Refresh plots with simulation_data.
+
+    Parameters
+    ----------
+    fig : Figure
+        A matplotlib.figure.Figure object.
+    ax : Axes
+        matplotlib.axes.Axes
+    single_simulation : dict
+        A single simulation
+    plot_type : str
+        Type of subplots
+    plot_config : dict
+        A dict specifies the preprocessing and style of plots.
+    """
+    # Make sure that visualization does not change the original data
+    dpls_copied = copy.deepcopy(single_simulation['dpls'])
+    net_copied = copy.deepcopy(single_simulation['net'])
+
+    for dpl in dpls_copied:
+        dpl.smooth(plot_config['dipole_smooth']).scale(
+            plot_config['dipole_scaling'])
+
+    if net_copied is None:
+        print("No network data")
+        return
+
+    if plot_type == 'spikes':
+        if net_copied.cell_response:
+            net_copied.cell_response.plot_spikes_raster(ax=ax, show=False)
+
+    elif plot_type == 'input histogram':
+        # BUG: got error here, need a better way to handle exception
+        if net_copied.cell_response:
+            net_copied.cell_response.plot_spikes_hist(ax=ax, show=False)
+
+    elif plot_type == 'PSD':
+        if len(dpls_copied) > 0:
+            dpls_copied[0].plot_psd(fmin=0, fmax=50, ax=ax, show=False)
+
+    elif plot_type == 'spectogram':
```

Comment:
Something seems to be going wrong with clearing the axis for this plot, when you hit the `Plot` button multiple times the colorbars keep getting added

Reply:
Indeed but it seems if we want to change the behavior here we need to modify `plot_tfr_morlet`. Could we leave this to later other PRs which may involve the change of other visualization functions also?

Reply:
Added to #497